### PR TITLE
Disallow `impl AsyncDrop` for unions

### PR DESCRIPTION
**Pull request by AndyGauge**
## What

Rejects `impl AsyncDrop for <union>` at coherence time. A union may still implement the synchronous `Drop` trait, but it has no async drop glue, so dropping a union in a synchronous context would silently skip any async cleanup.

The check is added to the existing `visit_implementation_of_drop` in `coherence/builtin.rs`, which already runs for both `Drop` and `AsyncDrop`. When the self type is a local union **and** the trait being implemented is `AsyncDrop`, we emit a new `AsyncDropOnUnion` diagnostic instead of accepting the impl.

## Why

Resolves `FIXME(zetanumbers): Disallow types with AsyncDrop in unions` and implements the *"union types rejecting non-trivially async destructible fields"* item on the async drop tracking issue ([rust-lang/rust#126482](https://github.com/rust-lang/rust/issues/126482)).

The field case is already closed transitively: `AsyncDrop` requires a `Drop` impl, `Drop` conflicts with `Copy`, and union fields must be `Copy`-like — so a bare async-drop field is impossible today. The remaining live hole was a **direct** `impl AsyncDrop` on a union, which previously compiled and ran its async destructor (demonstrated in the old `async-drop-initial` run-pass test).

## Safety / blast radius

- Gated behind the incomplete `async_drop` feature — cannot affect stable code.
- Only *adds a rejection*; the worst case is a false positive (a compile error), not unsoundness.
- Synchronous `Drop` on a union is unaffected and still compiles.

## Tests

- New `tests/ui/async-await/async-drop/union.rs` asserts the error (and that sync `Drop` on a union still compiles).
- The now-illegal `AsyncUnion` case is removed from `async-drop-initial.rs` / `.run.stdout` and its miri mirror (`src/tools/miri/tests/pass/async-drop.*`). The miri changes are **test fixtures only** — no interpreter code touched.
- Full `tests/ui/async-await/async-drop` suite: 31 passing locally (stage 1).
- Edge cases verified manually: generic unions caught; no double-error when sync `Drop` is absent; sync `Drop` on a union still allowed.

## Open question for review

Should the restriction reject the **impl** (as done here) or the **fields**, and is `coherence/builtin.rs` the right home for it? The tracking-issue wording is field-oriented, but the field case is already transitively closed, so rejecting the impl is what actually closes the hole. Happy to move it or add a field-level check if preferred.

---
🤖 This PR was implemented with Claude Code (LLM-assisted) under human direction. See the "[Async Drop](https://andygauge.github.io/async-drop/)" research-log book for a full disclosure and a line-by-line walkthrough of the change.

### CODE REVIEW

**Comment by P8L1** [comment]
You opened the PR against your own Rust fork instead of the official "rust-lang/rust" repository. Also, do not include a testing section.

**Comment by AndyGauge** [comment]
> ou opened the PR against your own Rust fork

Correct, I was trying to figure out if this was what was being asked by a TODO comment before re-writing it by hand and opening it against the official repository.  Given I never actually found who to ask, I think you are right that for feedback I need to just open a PR against r-l/r.  I was experimenting with how things could work with an LLM-assistance, and your feedback about the testing description section is spot on.  I will work towards re-writing this by hand and opening it up against the official during this week.  Thank you!

File: compiler/rustc_hir_analysis/src/coherence/builtin.rs [hunk 1]
```diff
@@ -77,7 +77,18 @@ fn visit_implementation_of_drop(checker: &Checker<'_>) -> Result<(), ErrorGuaran
     let impl_did = checker.impl_def_id;
     // Destructors only work on local ADT types.
     match checker.impl_header.trait_ref.instantiate_identity().skip_norm_wip().self_ty().kind() {
-        ty::Adt(def, _) if def.did().is_local() => return Ok(()),
+        ty::Adt(def, _) if def.did().is_local() => {
```
Comment (AndyGauge) [inline compiler/rustc_hir_analysis/src/coherence/builtin.rs:80]: **Critic:** I think this should be in `wfcheck` instead of `coherence`

**Judge (accept):** Well-formedness checks, such as validating that a type satisfies the structural requirements for implementing a specific trait (like `AsyncDrop`), belong in `wfcheck` rather than coherence, which is primarily responsible for orphan rules and overlap detection.

_(reviewer-27b@epoch1 critic → Qwen3.6-27B judge)_

File: tests/ui/async-await/async-drop/async-drop-initial.rs [hunk 2]
```diff
@@ -107,8 +107,6 @@ fn main() {
             48,
         )
         .await;
-
-        test_async_drop(AsyncUnion { signed: 21 }, 32).await;
```
Comment (AndyGauge) [inline tests/ui/async-await/async-drop/async-drop-initial.rs:111]: **Critic:** I'm not sure why this test was removed.

**Judge (accept):** The comment is valid because removing a test case without explanation or a corresponding issue reference risks regressing coverage for `AsyncUnion` drop behavior, so the author should clarify the rationale.

_(reviewer-27b@epoch1 critic → Qwen3.6-27B judge)_

File: tests/ui/async-await/async-drop/union.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    fn drop(&mut self) {}
+}
+
+// Rejected: a union may not implement `AsyncDrop`.
+impl AsyncDrop for U {
+    //~^ ERROR the `AsyncDrop` trait may not be implemented for unions
+    async fn drop(self: Pin<&mut Self>) {}
+}
+
+fn main() {}
```
Comment (AndyGauge) [inline tests/ui/async-await/async-drop/union.rs:33]: **Critic:** I think we should also have a test for a union with a field that has `AsyncDrop` implemented.

**Judge (accept):** The current test only verifies that a union cannot directly implement `AsyncDrop`, but it does not verify the more complex case where a union contains a field that implements `AsyncDrop`, which is also prohibited by the design and should be tested to ensure the compiler correctly rejects it.

_(reviewer-27b@epoch1 critic → Qwen3.6-27B judge)_

**Review by AndyGauge — COMMENTED** [review]
Automated review: `reviewer` LoRA (Qwen3.6-27B, checkpoint-1000-epoch1) critic findings, verdicted by a Qwen3.6-27B judge. 3 of 12 raw findings were judged ACCEPT; posting those here.

**Comment by AndyGauge** [comment]
See https://github.com/rust-lang/rust/pull/159254
